rust/src/write.rs

```
@@ -1053,14 +1051,38 @@ impl<W: Write + Seek> Writer<W> {
         ccw.write_u64::<LE>(summary_start)?;
         ccw.write_u64::<LE>(summary_offset_start)?;
 
-        let (mut writer, summary_crc) = ccw.finalize();
+        let (writer, summary_crc) = ccw.finalize();
 
         writer.write_u32::<LE>(summary_crc.finalize())?;
 
         writer.write_all(MAGIC)?;
         writer.flush()?;
         Ok(())
     }
+
+    /// Consumes this writer, returning the underlying stream. Unless [`Self::finish()`] was called
+    /// first, the underlying stream __will not contain a complete MCAP.__
+    ///
+    /// Use this if you wish to handle any errors returned when the underlying stream is closed. In
+    /// particular, if using [`std::fs::File`], you may wish to call [`std::fs::File::sync_all()`]
+    /// to ensure all data was sent to the filesystem.
+    pub fn into_inner(mut self) -> W {
+        self.is_finished = true;
+        match self.writer.take().expect(Self::WHERE_WRITER) {
+            WriteMode::Raw(w) => w.finalize().0,
+            WriteMode::Attachment(w) => w.writer.finalize().0.finalize().0,
+            WriteMode::Chunk(w) => {
+                w.compressor
+                    .finalize()
+                    .0
+                    .finish()
+                    .expect("compression error")
```

Did you mean to have this here instead of returning a result?

---

Yes. It's because of the zstd compressor. I'll add a comment to clarify.

---

what if we swallowed the error instead? The caller is already expecting to not get a complete MCAP.

---

I've rejigged it to avoid the unwrap entirely by using the interface in zio::Writer instead.

---

@james-rms swallowing wasn't an option because the wrapper interface wouldn't even let you get the W back unless finishing succeeded.